includes/class-wc-payments-payment-request-button-handler.php

```diff
@@ -496,6 +512,33 @@ public function has_allowed_items_in_cart() {
 		return true;
 	}
 
+	/**
+	 * Checks whether cart contains a subscription product or this is a subscription product page.
+	 *
+	 * @return boolean
+	 */
+	public function has_subscription_product() {
+		if ( ! class_exists( 'WC_Subscriptions_Product' ) ) {
+			return false;
+		}
+
+		if ( $this->is_product() ) {
+			$product = $this->get_product();
+			if ( WC_Subscriptions_Product::is_subscription( $product ) ) {
+				return true;
+			}
+		} elseif ( $this->is_checkout() || $this->is_cart() ) {
+			foreach ( WC()->cart->get_cart() as $cart_item_key => $cart_item ) {
+				$_product = apply_filters( 'woocommerce_cart_item_product', $cart_item['data'], $cart_item, $cart_item_key );
+				if ( WC_Subscriptions_Product::is_subscription( $_product ) ) {
+					return true;
+				}
```

Comment:
So, If I'm reading this correctly we will use this method to tell the user they need to log in if at least one subscription is present on the cart right? In that case are we aware the user will not be able to purchase other products if a subscription is present in the cart? My point is, we are not specifically telling the user they can not purchase the other products due to a subscription being present in the cart.

The message in the cart doesn't explicitly tells the user the subscription is the one that requires the user to sign in to be able to purchase the products.

Reply:
> if at least one subscription is present on the cart right?

Exactly, and if it's not possible to create an account upon checkout.

> The message in the cart doesn't explicitly tells the user the subscription is the one that requires the user to sign in to be able to purchase all products.

I believe we should view this as: "Can I buy the items on the cart without having an account?" If not, I should be redirected to the login page. I guess if we say something about a specific product being the blocker, we could implicitly tell people to remove the subscription in order to proceed with the order. 🤔

We don't want users to give up on a subscription from their cart just because they need to create an account first, so in my opinion the message is fine.

I didn't personally craft the messaging, but we can ask for a design opinion considering this scenario. Do you think we should?

Reply:
> "Can I buy the items on the cart without having an account?" If not, I should be redirected to the login page.

I see what you mean now. If the copy was made considering the case I'm presenting then looks good to me.

Reply:
Just confirmed with David and the message covers both cases. Marking as resolved ✅